docs/design-patterns/1.creation-pattern/singleton.md

```diff
@@ -0,0 +1,136 @@
+
+## Lazy singleton
+
+Suppose the following cases:
+
+1. The object requires some private variables and private methods.
+2. The object itself is quite complicated, so it will take additional time to create before usage.
+
+So, it is no longer possible to use object literals to create singleton. Instead, we need to instantiate the object as a constructor.
+
+### ES5
+
+Here's how to transform the above `foo` object by **IIFE** and constructor in ES5 syntax.
+
+```js
+let Singleton = (function(name) {
+  let instance = null;
+  
+  function init() {
+    this.name = name,
+    this.doSomething = function() {} // method
+  }
+  
+  return function() {
+    if(!instance) {
+      instance = new init();
+    }
+    return instance;
+  }
+})()
+```
+
+The above `Singleton` is actually a **IIFE**, `instance` as the instance is initially assigned to `null`, `init` is actually a constructor, used to instantiate the object, the immediate execution of the function returns an anonymous function to determine whether the instance is created, only when `Singleton()` is called to instantiate the instance.
+
+Because the return of the **IIFE** is a function, the instance is created only when called. This is called **lazy singleton**. Instead of creating instance when loaded, the instance will be created if needed. The first instantiated instance object is always returned if called again.
+
+```js
+let instanceA = Singleton();
+let instanceB = Singleton();
+console.log(instanceA === instanceB); //true
+```
+
+In addition, the singleton in JS usually is to manage shared application state, such as using a singleton as the source of config settings for a web app on the client side for anything initiated with an API key, storing data in memory in a client-side web application (e.g. stores in Redux). So, `Singleton` can be seen as `Store` and `instance` as shared `data` to some extent.
+
+### ES6
+
+```js
+class Singleton {
+  constructor(name) {
+    this.name = name;
+  }
+  // method
+  doSomething () {}
+  //static method
+  static getInstance(name) {
+    if(!this.instance) {
+      this.instance = new Singleton(name);
+    }
+    return this.instance;
+  }
+}
+```
+
+```js
+let instanceA = Singleton.getInstance('');
+let instanceB = Singleton.getInstance('')
+
+console.log(instanceA === instanceB); //true
+```
+
+## Singleton in React
+
+We should keep backdrop single when the aside is nested. See below example:
```

Comment:
I think this content is abstract, because I read this content several times before I understood it.